File: snippets/ruby/sort_bubble_sort.rb
def do_sorting(input_list)
  bubble_sort(input_list)
end

def bubble_sort(input_list)
  output_list = input_list
  is_sorted = false

  while is_sorted == false
    output_list, is_sorted = do_sorting_round(output_list)
  end

  output_list
end

def do_sorting_round(input_list)
  output_list = []
  is_sorted = true

  input_list.each_with_index do |element, index|
    if index.zero?
      output_list.append(element)
    elsif element < output_list[index - 1]
      previous_element = output_list[index - 1]
      output_list.pop
      output_list.append(element)
      output_list.append(previous_element)
      is_sorted = false
    else
      output_list.append(element)
    end
  end

  [output_list, is_sorted]
end

File: snippets/rust/sort_bubble_sort.rs
fn do_sorting(input_list: Vec<&str>) -> Vec<&str> {
    return bubble_sort(input_list);
}

fn bubble_sort(input_list: Vec<&str>) -> Vec<&str> {
    let mut output_list = input_list;
    let mut is_sorted = false;

    while is_sorted == false {
        let (_output_list, _is_sorted) = do_sorting_round(output_list);
        output_list = _output_list;
        is_sorted = _is_sorted;
    }

    return output_list;
}

fn do_sorting_round(input_list: Vec<&str>) -> (Vec<&str>, bool) {
    let mut output_list: Vec<&str> = vec![];
    let mut is_sorted = true;

    for (index, element) in input_list.iter().enumerate() {
        if index == 0 {
            output_list.push(element);
        } else if element < &output_list[index - 1] {
            let previous_element = output_list[index - 1];
            output_list.pop();
            output_list.push(element);
            output_list.push(previous_element);
            is_sorted = false
        } else {
            output_list.push(element);
        }
    }

    return (output_list, is_sorted);
}
